objects/src/transaction/tx_result.rs

```diff
@@ -0,0 +1,171 @@
+use super::{
+    AccountDelta, AccountId, AccountStub, AdviceProvider, ConsumedNotes, CreatedNotes, Digest,
+    Program, StackOutputs, TransactionResultError, TransactionWitness, TryFromVmResult,
+};
+use crypto::utils::collections::Diff;
+use miden_processor::RecAdviceProvider;
+
+/// [TransactionResult] represents the result of the execution of the transaction kernel.
+///
+/// [TransactionResult] is a container for the following data:
+/// - initial_account_stub: a stub of the account before the transaction was executed.
+/// - final_account_stub: a stub of the account after the transaction was executed.
+/// - account_delta: a delta between the initial and final account stubs.
+/// - consumed_notes: the notes consumed by the transaction.
+/// - created_notes: the notes created by the transaction.
+/// - block_hash: the hash of the block against which the transaction was executed.
+/// - program: the program that was executed.
+/// - tx_script_root: the script root of the transaction.
+/// - advice_provider: the final state of the advice provider.
+/// - stack_outputs: the stack outputs produced by the transaction.
+#[derive(Debug, Clone)]
+pub struct TransactionResult<T: AdviceProvider> {
+    initial_account_stub: AccountStub,
+    final_account_stub: AccountStub,
+    account_delta: AccountDelta,
+    consumed_notes: ConsumedNotes,
+    created_notes: CreatedNotes,
+    block_hash: Digest,
+    program: Program,
+    tx_script_root: Option<Digest>,
+    advice_provider: T,
+    stack_outputs: StackOutputs,
+}
```

Comment:
Not a review yet, but just a few questions about this struct:

1. Why do we need to have `AccountStub`'s here for initial/final state rather than just commitments to the account (i.e., `Digest'`s)?
2. Why do we need `stack_outputs` here? Shouldn't all data in stack outputs be contained in the other fields here? Or am I forgetting something?
3. Why do we put the full advice provider here? I was thinking that by this point we'd already extract all the relevant data from the advice provider.

Overall, I was thinking of the `TransactionResult` as an object which contains:
1. New data generated during transaction execution (primarily `account_delta` and `created_notes`).
2. Minimal set of data that we need to re-execute the transactions (i.e., data that we need to build transaction witness).

But it seems like currently it contains quite a bit more info than that. Or am I missing something?

Reply:
Yes we can definitely reduce data duplication / overhead here, I will go head and do so.

The main open question here is around how we compute the account_delta.  My thinking is that we pass the initial_account: `Account` as an argument to the constructor of `TransactionResult`. Within the `TransactionResult` constructor we extract the final_account: `Account` from the `StackOutputs` and `AdviceProvider`.  We then perform a initial_account.diff(&final_account) to produce the `AccountDelta`.  To support this pattern we need to be able to implement `TryFromVmResult` for `Account` to extract the final_account.  One of the challenges here is we have some data structures that we need to extract - specifically: 

```rust
pub struct AccountCode {
    module: ModuleAst,
    procedures: Vec<Digest>,
    procedure_tree: SimpleSmt,
}
```

```rust
pub struct AccountStorage {
    slots: SimpleSmt,
    store: MerkleStore,
}
```

```rust
pub struct AccountVault {
    asset_tree: TieredSmt,
}
```

Maybe its fine to assume that the `AccountCode` is unchanged and we can clone it from `initial_account` for now (this should be fine until we introduce the transaction type that allows for account code to be changed). However in the future we will need to understand how we will approach this.  

For the `AccountStorage` we need to be able to extract both a `SimpleSmt` and a `MerkleStore` subset from the `MerkleStore`.  I think this is practical now, maybe we introduce a trait

```rust
trait TryFromMerkleStore: Sized {
    type Error;
    
    fn try_from_merkle_store(root: Digest, store: &MerkleStore) -> Result<Self, Error>;
}
```

and implement this for `SimpleSmt`. The store: `MerkleStore` can easily be extracted using the `.subset(..)` method.

We would also need to implement `TryFromMerkleStore` for `TieredSmt` however I'm not sure how practical that is at the minute?

We would then implement `Diff` and `ApplyDiff` on `Account` (and potentially on `AccountStorage`, `AccountCode`, `AccountVault`) to compute and apply diffs.

Reply:
Here is my high-level thinking on this (I might be missing something as I haven't spent too much time thinking about this yet).

First, I agree, let's put aside account code for now as we can leave it for future PRs. Second, the route of computing the diff between two account objects may be more complicated than what we need in this case. I think we could build a specialize procedure which can compute the delta based on the data in the Merkle store and stack inputs/outputs.

### Account storage
The first thing we want to get is the list of leaves for `slots` which have changed during transaction execution (we could get a full `SimpleSmt` - but I'm not sure that's necessary). To get these, we could add a specialized method to `RecAdviceProvider`. This method would traverse `MerkleStore` from a given root to a given depth (we would need to a method to `MerkleStore` for this), and then would filter out nodes which haven't changed based on the information in `RecordingMap`'s `updates` set.

Once we have a list of `slots` which were updated, we could again traverse the `MerkleStore` in the `RecAdviceProvider` for each of them to:

1. Check if the new slot value is a root of a Merkle tree in the `MerkleStore`.
2. Get updated leaves for this Merkle tree.

Not sure yet what's the best way to do the latter. We might have to build a custom iterator for `MerkleStore` which iterates over leaves of a Merkle tree with a given root.

Basically, the approach that I would try to take is to add specialized methods to `RecAdviceProvider` to help with extracting delta from there because:

1. I think most of the info that we need is already there, and we should be able to use this info to extract the delta relatively efficiently.
2. This functionality may be useful beyond the rollup.

### Account vault
We could use the similar approach as described above to get leaves of the vault which were modified (and it is actually going to be a bit simpler as we only have a single root rather than a set of potential roots).

But the extra complication here is that we also need to convert each leaf into a key-value pair. We can do this by looking up leaf pre-images in the advice map. One issue is that we'll probably end up with some "false positives" here (i.e., key-value pairs which are actually old, but appear as new because their location in the tree has changed) - but I think we can solve this problem later.

Reply:
> First, I agree, let's put aside account code for now as we can leave it for future PRs. Second, the route of computing the diff between two account objects may be more complicated than what we need in this case. I think we could build a specialize procedure which can compute the delta based on the data in the Merkle store and stack inputs/outputs.

Yes I think this is practical and can be done.

> ### Account storage
> The first thing we want to get is the list of leaves for `slots` which have changed during transaction execution (we could get a full `SimpleSmt` - but I'm not sure that's necessary). To get these, we could add a specialized method to `RecAdviceProvider`. This method would traverse `MerkleStore` from a given root to a given depth (we would need to a method to `MerkleStore` for this), and then would filter out nodes which haven't changed based on the information in `RecordingMap`'s `updates` set.

I'm not sure if this will work as the `RecordingMap`'s`updates` set is global across all Merkle structures. A node may be in the `updates` set because it was added to a child tree (or some other Merkle strucutre). An alternative would be to introduce a `diff_merkle_tree(root_1, root_2, merkle_config)` method on the `AdviceProvider` which traverses the tree and returns a diff object:

```rust
pub struct MerkleDelta {
    cleared_slots: Vec<u8>,
    updated_slots: Vec<(u8, Word)>,
}
```

This would work for all types of AdviceProvider. This is done by traversing the trees to the leaves and identifying leaves which are different between the two trees.  What do you think?

> Once we have a list of `slots` which were updated, we could again traverse the `MerkleStore` in the `RecAdviceProvider` for each of them to:
> 
> 1. Check if the new slot value is a root of a Merkle tree in the `MerkleStore`.
> 2. Get updated leaves for this Merkle tree.
> 
> Not sure yet what's the best way to do the latter. We might have to build a custom iterator for `MerkleStore` which iterates over leaves of a Merkle tree with a given root.

One of the challenges we have here is that the Merkle structures in the `MerkleStore` are "weakly typed", as such we don't know the structure (`SimpleSmt` / `TieredSmt` / `Mmr`) and its properties (depth, tier depths etc).  This may make the process of extraction somewhat more challenging. Currently we are extract the child Merkle structures into initial and final `MerkleStore`s and then perform a diff to create a `KvMapDiff`.  This `KvMapDiff` is less efficient as it also contains the inner node diffs however it does not require us to know type information. It also supports infinitely nested structures i.e. child trees below child trees below child trees. However, I suspect this will be more challenging if we want to support `TieredSmt` which requires us to extract values from the advice map. 

https://github.com/0xPolygonMiden/miden-base/blob/71b84569d2c7f8da52430bbe24a25f67851a4dfd/objects/src/accounts/storage.rs#L144-L146

What do you think?

> Basically, the approach that I would try to take is to add specialized methods to `RecAdviceProvider` to help with extracting delta from there because:
> 
> 1. I think most of the info that we need is already there, and we should be able to use this info to extract the delta relatively efficiently.
> 2. This functionality may be useful beyond the rollup.

This makes sense.

> ### Account vault
> We could use the similar approach as described above to get leaves of the vault which were modified (and it is actually going to be a bit simpler as we only have a single root rather than a set of potential roots).
> 
> But the extra complication here is that we also need to convert each leaf into a key-value pair. We can do this by looking up leaf pre-images in the advice map. One issue is that we'll probably end up with some "false positives" here (i.e., key-value pairs which are actually old, but appear as new because their location in the tree has changed) - but I think we can solve this problem later.

Sounds good.

Reply:
> I'm not sure if this will work as the `RecordingMap`'s`updates` set is global across all Merkle structures. A node may be in the `updates` set because it was added to a child tree (or some other Merkle strucutre). An alternative would be to introduce a `diff_merkle_tree(root_1, root_2, merkle_config)` method on the `AdviceProvider` which traverses the tree and returns a diff object:
> 
> ```rust
> pub struct MerkleDelta {
>     cleared_slots: Vec<u8>,
>     updated_slots: Vec<(u8, Word)>,
> }
> ```
> 
> This would work for all types of AdviceProvider. This is done by traversing the trees to the leaves and identifying leaves which are different between the two trees. What do you think?

This makes sense, though I'm a bit hesitant to add this to the general `AdviceProvider` interface. The reason is that we have some contextual knowledge here (i.e., both trees are fully-balanced and have depth 8) and without this contextual knowledge things become a bit more complicated.

There is another challenge: we need to extract this information before `RecAdviceProvider` is converted into a proof, but accessing data in `RecAdviceProvider` (at least using regular methods) will modify the proof itself. So, we might want to change how finalization of `RecAdviceProvider` happens (e.g., maybe the finalizer should return a tuple `(AdviceInputs, MerkleStore, BTreeMap)` and then we can extract required info from the returned store and advice map).

For `MerkleStore`, I think we could add a method which iterates over nodes at a given depth for a tree with a given root, and then use this method to build the `MerkleDelta` struct you described above.

> One of the challenges we have here is that the Merkle structures in the `MerkleStore` are "weakly typed", as such we don't know the structure (`SimpleSmt` / `TieredSmt` / `Mmr`) and its properties (depth, tier depths etc). This may make the process of extraction somewhat more challenging. Currently we are extract the child Merkle structures into initial and final `MerkleStore`s and then perform a diff to create a `KvMapDiff`. This `KvMapDiff` is less efficient as it also contains the inner node diffs however it does not require us to know type information. It also supports infinitely nested structures i.e. child trees below child trees below child trees. However, I suspect this will be more challenging if we want to support `TieredSmt` which requires us to extract values from the advice map.

I think we might have to impose some type restrictions as supporting fully generic structures may be too complicated. For example, a slot may contain a preimage for data which is located in an advice map. Or maybe this data in the advice map actually represents peaks of MMR which are also in the advice provider. And maybe this MMR has some nested trees in it roots of which also need to be looked up in the advice map.

So, my thinking at this point is that we start very simple and extend the functionality later. The progression could look like this:

1. Storage slots are just values.
2. Storage slots may contain roots of simple SMTs.
3. Storage slots may contain preimages of sequential hashes (located in the advice map).
4. Storage slots may contain tiered SMTs.
5. Maybe something else.

For this PR, maybe we do just 1 and 2 (or maybe even just 1 and add 2 in a subsequent PR).